### DISCUSSION TITLE
How do I make my RPi All Sky camera publicly available?

### DISCUSSION THREAD

**Discussion by gigaklatch** [aaronwmorris/indi-allsky#943] [Q&A]
Hi there,

I am not at all well-versed in stuff like these. So, how on earth do I make it happen?

I have a Raspberry Pi 4 running indi-allsky at a remote location , there is accesss to the internet via 4G. I would like to have the ability to access the Web Interface of the  from wherever I am myself, just like if I was on the same network.

Using the VNC software that is included in the RPi OS I can connect directly to the Pi and look through the web interface locally on the RPi.  But this basically is an interactive video stream with quite poor quality, and that is not what I am after.

It seems like one could simply tell the mobile 4G router that the RPi is connected to to port forward anyone coming and knocking on a specific port. But that is quite unsafe from what I have gathered.

Should I use SyncAPI? But how and where do I run that remote web server instance of indi-allsky? 

Please help

**Comment by aaronwmorris** [comment]
The indi-allsky web interface was designed to be directly exposed to the Internet.  If you can get the port forward setup, that should work quite well.  You may have to use ipv6, but that should not be a problem.

**Reply by gigaklatch** [reply]
Ah okay, that's good! Thank you for the quick response.

Another problem though is that the IP-address for the 4G connection is neither public nor static. So getting around that will take some additional engineering.  Know of any ways other people in similar situations have solved that?

**Reply by bluthen** [reply]
One way is you have a server that does have a public ip address and you create a vpn. Then on the public machine you proxy or forward from the public ip address to the pi's vpn ip address. You can buy a minimal cloud system to be the VPN, and let your internal machine do the real serving.

Your own internal VPN is useful for lots of other things, so it is something I set up anyway. 

If your 4G connection has limited data transfer or it has high data costs, you might not want to ever make it publicly accessible. Simply connecting to it through the vpn connection for yourself only when on the vpn might be all you want to do. 

If you want other people to see it, then the syncapi is a better option as you can have something with less data costs be the public front.

**Reply by aaronwmorris** [reply]
If the IP is not public, but a NAT address, obviously you will not be able to access it directly.

The SyncAPI is an option for you.  You would have to run a separate installation of indi-allsky on a publicly available cloud server.  The SyncAPI would sync the images and videos to the cloud server so you could view them.  You would not be able to directly control the camera, though.  The SyncAPI is mainly to make the images and videos viewable remotely.

**Reply by Vertikar** [reply]
Tailscale for VPN could be a good option to be able to access and manage it, and like bluthen and aaron have suggested, SyncAPI might be the best option to make your images and timelapses publicly available

**Comment by gigaklatch** [comment]
I decided to try having indi-allsky run on a Google Cloud VM.  Typing in the IP-address for the cloud VM in a browser at home I can reach the indi-allsky web interface of the VM, so it seems to be working, at least to some degree.

Is the "foobar123" in the instructions for setting up synciAPi a placeholder for where I can enter my own username? Or should I go with the "foobar123"?

On the "local" machine I get the following message( I have edited out the IP-address to my cloud VM):

",requests_syncapi_v1 file transfer connection failure: HTTPSConnectionPool(host='SUPER-SECRET-IP-ADDRESS', port=443): Max retries exceeded with url: /indi-allsky/sync/v1/camera (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1123)')))"

Does this tell you what I am doing wrong and need to do differently??

**Reply by aaronwmorris** [reply]
You can use any username you wish to setup.

Since the setup is using a self-signed certificate by default, you will need to disable certificate validation for SyncAPI.

**Comment by gigaklatch** [comment]
Thank for all the great pointers! 

Disabling the certificate validation did the trick. At least the images were being synced but I did not wait for it being time for a timelapse to get synced, so I do not know if that worked or not. 

Instead, I decided to also try out the suggestion of Tailscale. And man.. I am loving that little piece of software! Super easy to use and quick! As I have a lil' NUC always on at home  I installed indiallsky and tailscale on a VM running on it. So both the VM on the NUC and the RPi peering up at the night sky are in the same tailscale tunnel. The syncing of the images between my indiallsky devices works perfectly. Unfortunately though, for the timelapses this is not the case...  

Last night, soon after the daytime timelapse had finished brewing I had a message flashing by saying something about something timing out or whatever. Is there somewhere I can find some kind of log and perhaps see what goes wrong when the Pi is trying to upload the timelapse videos?

**Reply by aaronwmorris** [reply]
The normal log at `/var/log/indi-allsky/indi-allsky.log` will have all of the log entries.  You can grep for the `MainProcess/Upload`  threads to just see upload information.

Individual files and videos upload status are tracked in the database, so you may synchronize any missing files with the `./misc/upload_sync.py` utility.

**Comment by gigaklatch** [comment]
- Right before the upload starts:

Sep 18 06:01:40 raspberrypi [WARNING] Video-5/MainThread miscUpload.upload_video() #93: Video uploading disabled

- Right after this, it seems to start the upload anyway? ( with details such as IP adress and user name altered for privacy):

Sep 18 06:01:40 raspberrypi [INFO] MainProcess/Upload-9 generic.connect() #40: Connecting to https://SUPERSECRETIP/indi-allsky/sync/v1/video (443) as MR_USER with requests_syncapi_v1

- Six seconds later:

Sep 18 06:01:46 raspberrypi [ERROR] MainProcess/Upload-9 uploader.processUpload() #369: Connection failure: ('Connection aborted.', timeout('The write operation timed out'))

The same thing repeats some minutes later for the strail video.

Meanwhile, the images keep syncing just fine.

The upload speed of the connection is about 1 megabyte per second. Doing a manual download of any timelapses, or videos, by right-clicking and pressing "Save link as..." works fine and dandy and takes about two minutes. 

The section of the wiki for SyncAPI-Setup mentions enabling multiple upload workers. I do not see such a setting under the  SyncAPI tab, but under the file transfer there is a settings like that. which is set to "2".  Does this affect the SyncAPI too? Should I try increasing this? Could doing all of this over a 4G connection an issue?

**Reply by aaronwmorris** [reply]
Well, I have the timeout for the syncapi connection set to 5 seconds.  I had assumed that would only affect the connection timeout, but maybe its for the entire upload process.  The images sync fine because they are small.  I will add a custom timeout config for the syncapi.

As for the upload workers, all file transfers, mqtt, syncapi all use the same workers, so that is the setting for increasing/decreasing the workers.  For a slower connection, you may want to reduce this to 1 worker.

**Reply by aaronwmorris** [reply]
Merged #953 with the new timeout setting for the syncapi

**Reply by gigaklatch** [reply]
Splendid!

How does the timeout work though? Will  the timeout trigger and abort any uploads that are still ongoing as soon as the set amount of seconds have passed?

There is a great risk of the upload taking longer than 120 seconds in my case, with a max upload speed of 1 megabyte per seconds and surely lower than that a lof ot the time(mobile connection). My night timelapses are about 120 megabytes at the moment, and they will get bigger as the nights become longer and longer.

**Reply by aaronwmorris** [reply]
The timeout is supposed to be for both connecting and responses, not for the full transfer.  In your case, I think the slow connection is just causing slow responses from the server.  You may only need to increase the timeout to 10 or 15 seconds to fix your situation.

**Comment by gigaklatch** [comment]
I have only tried up to 30 seconds so far, but the only thing that seem to have changed is that it takes the set amount of before it timeouts just like before.

**Reply by aaronwmorris** [reply]
Make sure you are performing a config reload when you change the value.  It will not take effect unless you do.

**Reply by gigaklatch** [reply]
Hello,

I have had the config reload checked. I have rebooted the Pi several times. It seems  like the settings does stick as the time between when the upload start is logged and when the timeout error is logged changes with the timeout value set in config.

Now that the nights are longer the videos get a little bit bigger, and therefore takes that little bit longer to upload to the remote instance of Indi Allsky, none of the timelapse uploads are succesful.

I need to get myself some faster internet it seems like.

**Reply by aaronwmorris** [reply]
I have increased the max timeout to 20 minutes in #959
